test/cqlpy/test_map_key_selection.py

```diff
@@ -0,0 +1,63 @@
+# Copyright 2020-present ScyllaDB
+#
+# SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+
+#############################################################################
+# Tests for ...
```

Comment:
Obviously you'll need to change this line ;-)

Also, please note that the Cassandra feature (see https://github.com/scylladb/scylladb/issues/7751) also supports selecting elements from sets, not just maps. You should probably implement both (why do only half?), so this test file should have a name like test_select_collection_element - not "map_key_selection".

Reply:
I'd do set in a different PR, just because we have an urgent request for a selection from a map.

Reply:
@mykaul I don't agree because it seems that by some miracle (probably @avikivity 's foresight and not a miracle) set is already supported (as @mlitvk 's test of `frozen<map<text, set<int>>>` demonstrates) - so we better notice that it's supported, spend a few more minutes to verify this issue, and close that entire Cassandra-4.0 incompatibility issue instead of half of it.

In any case, even if you really really don't want to finish the set coding/testing today, the minimum you can do is to call a test file which you **know** will also host tests for sets, not just maps, with a suitable name - use the word collection instead of sets. This is even more true given that the test already contains one test with a set in it.

Reply:
It shouldn't work. evaluate() for a subscript boils down to

```c++
managed_bytes_opt
get_value(const subscript& s, const evaluation_inputs& inputs) {
    raw_value container_val = evaluate(s.val, inputs);
    auto serialized = std::move(container_val).to_managed_bytes_opt();
    if (!serialized) {
        // For null[i] we return null.
        return std::nullopt;
    }
    auto col_type = static_pointer_cast<const collection_type_impl>(type_of(s.val));
    const auto deserialized = type_of(s.val)->deserialize(managed_bytes_view(*serialized));
    const auto key = evaluate(s.sub, inputs);
    auto&& key_type = col_type->is_map() ? col_type->name_comparator() : int32_type;
    if (key.is_null()) {
 ...
   }
    if (col_type->is_map()) {
...
    } else if (col_type->is_list()) {
...
    } else {
        throw exceptions::invalid_request_exception(fmt::format("subscripting non-map, non-list column {:user}", s.val));
    }
}
```

So, we should see an exception. It would also be surprising if prepare works.


I don't think we should support `SELECT set[key]` since it returns the key itself (or NULL)! Rather we should support `SELECT key IN set` instead (returning a boolean).

Reply:
yes, the subscript works also for lists, but not for sets (cassandra doesn't support list subscripts)

Reply:
Then claim list support in changelog/subject.

Reply:
As I noted elsewhere, Cassandra does support this syntax for sets (IIUC returns true/false depending if a given set member exists) so ideally we should to unless it's very difficult to do (and if it is, we should document this difference). Cassandra decided not to support list indexing (the subscript is an integer) because it is inefficient - if we do support we need to test it and document it as a Scylla-only feature. 

We also need to be really careful about or extensions - because Cassandra may suddenly decide to support a missing feature in a different way from what we implemented, and then we'll be stuck between a rock and hard place. We can seriously consider artificially refusing list subscripts, explaining in a comment that although we could support it, we decided not to because Cassandra doesn't.

Reply:
we already support list subscripts in UPDATE where cassandra doesn't, so I can't disable it entirely.
I'm not sure if I should disable it now just in SELECT

Reply:
Ok, then I guess we'll need to document this difference - I see the "WHERE" part is already documented in docs/cql/cql-extensions.md so we'll need to also mention the ability to select an index. And we'll need to have write regression tests for it, so it doesn't break in the future without anyone noticing (and since the translated Cassandra tests don't cover this case at all).

Reply:
I added support of sets

Reply:
another difference I see with Cassandra is in column conditions they allow only map and list, not sets.
For us it would work naturally when adding support of set subscripts because it's the same code for selectors and column conditions, while in Cassandra it's separate.
It allows to check for element existence in a set:
```
UPDATE tab SET s=s+{2} WHERE k=1 IF s[1]=1;
```
While the syntax is weird it could be useful and it's consistent with the other uses of set subscripts.

Reply:
Ok, though I really wish Cassandra was more careful with its syntax.